File: flutter/ios/Classes/DitoSdkPlugin.swift
import DitoSDK
import FirebaseCore
import FirebaseMessaging
import Flutter
import UIKit
import UserNotifications

public class DitoSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private static let notificationEventsChannelName = "br.com.dito/notification_events"
  private static let notificationClickEventType = "notification_click"
  private static var notificationEventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "br.com.dito/dito_sdk", binaryMessenger: registrar.messenger())
    let instance = DitoSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
    let eventsChannel = FlutterEventChannel(
      name: notificationEventsChannelName,
      binaryMessenger: registrar.messenger()
    )
    eventsChannel.setStreamHandler(instance)
    if FirebaseApp.app() == nil {
      FirebaseApp.configure()
    }
    DitoNotificationDelegate.ensureDitoConfigured()
    DitoNotificationDelegate.shared.configurePush(application: UIApplication.shared)
  }

  private static func isDitoChannel(_ userInfo: [AnyHashable: Any]) -> Bool {
    DitoNotificationDelegate.isDitoChannel(userInfo)
  }

  private static func resolvedFcmToken(_ fcmToken: String?) -> String {
    if let token = fcmToken?.trimmingCharacters(in: .whitespacesAndNewlines), !token.isEmpty {
      return token
    }
    return UserDefaults.standard.string(forKey: "FCMToken") ?? ""
  }

  private static func processNotificationReceived(userInfo: [AnyHashable: Any], fcmToken: String?) {
    DitoNotificationDelegate.ensureDitoConfigured()
    let normalized = DitoNotificationDelegate.normalizedDitoUserInfo(userInfo)
    Dito.notificationReceived(userInfo: normalized, token: resolvedFcmToken(fcmToken))
  }

  @objc public static func didReceiveNotificationRequest(
    _ request: UNNotificationRequest,
    fcmToken: String?
  ) -> Bool {
    let normalized = DitoNotificationDelegate.normalizedDitoUserInfo(request.content.userInfo)
    guard isDitoChannel(normalized) else { return false }
    processNotificationReceived(userInfo: request.content.userInfo, fcmToken: fcmToken)
    return true
  }

  @objc public static func didReceiveRemoteNotification(
    userInfo: [AnyHashable: Any],
    fcmToken: String?
  ) -> Bool {
    let normalized = DitoNotificationDelegate.normalizedDitoUserInfo(userInfo)
    guard isDitoChannel(normalized) else { return false }
    processNotificationReceived(userInfo: userInfo, fcmToken: fcmToken)
    return true
  }

  @objc public static func didReceiveNotificationClick(
    userInfo: [AnyHashable: Any],
    callback: ((String) -> Void)? = nil
  ) -> Bool {
    didReceiveNotificationClick(userInfo: userInfo, actionIdentifier: nil, callback: callback)
  }

  /// Same as above, but carries `UNNotificationResponse.actionIdentifier` so a tap on
  /// an action button reports which button was tapped.
  ///
  /// Kept as a separate entry point rather than a defaulted parameter on the one
  /// above, so the existing Objective-C selector stays untouched for host apps that
  /// already call it from an AppDelegate.
  @objc(didReceiveNotificationClickWithUserInfo:actionIdentifier:callback:)
  @discardableResult
  public static func didReceiveNotificationClick(
    userInfo: [AnyHashable: Any],
    actionIdentifier: String?,
    callback: ((String) -> Void)? = nil
  ) -> Bool {
    let normalized = DitoNotificationDelegate.normalizedDitoUserInfo(userInfo)
    guard isDitoChannel(normalized) else { return false }
    let received = Dito.notificationClick(
      userInfo: normalized,
      actionIdentifier: actionIdentifier,
      callback: callback
    )
    emitNotificationClickEvent(userInfo: normalized, received: received)
    return true
  }

  /// Emits the click on the Dart stream.
  ///
  /// The deeplink comes from `received.resolvedLink`, which is the button's own
  /// already-iOS-resolved link for an action tap and the notification's deeplink
  /// for a tap on the body. Everything else still comes from the raw payload:
  /// `DitoNotificationReceived` exposes the rich-push fields publicly but keeps
  /// `notification`, `reference`, `logId`, `notificationName` and `userId` internal,
  /// so they are not readable from this module.
  internal static func emitNotificationClickEvent(
    userInfo: [AnyHashable: Any],
    received: DitoNotificationReceived
  ) {
    guard let sink = notificationEventSink else { return }
    let normalized = DitoNotificationDelegate.normalizedDitoUserInfo(userInfo)
    let source = DitoNotificationDelegate.nestedPayload(normalized[AnyHashable("data")] ?? normalized["data"]) ?? normalized

    var payload: [String: Any] = [:]
    payload["type"] = notificationClickEventType
    payload["deeplink"] = received.resolvedLink
    payload["notificationId"] = source["notification"] as? String ?? ""
    // Sempre vazio: `reference` está em retirada e a atribuição ancora em
    // `user_id`. A chave permanece para não quebrar quem já lê este evento.
    payload["reference"] = ""
    payload["logId"] = source["log_id"] as? String ?? ""
    payload["notificationName"] = source["notification_name"] as? String ?? ""
    payload["userId"] = source["user_id"] as? String ?? ""
    payload["actionId"] = received.actionId
    payload["actionLabel"] = received.actionLabel
    payload["customData"] = received.customData

    DispatchQueue.main.async {
      sink(payload)
    }
  }

  public func application(
    _ application: UIApplication,
    didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data
  ) {
    Messaging.messaging().apnsToken = deviceToken
    Messaging.messaging().token { token, _ in
      guard let token, !token.isEmpty else { return }
      UserDefaults.standard.set(token, forKey: "FCMToken")
    }
  }

  public func application(
    _ application: UIApplication,
    didReceiveRemoteNotification userInfo: [AnyHashable: Any],
    fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void
  ) -> Bool {
    let normalized = DitoNotificationDelegate.normalizedDitoUserInfo(userInfo)
    let isDito = DitoSdkPlugin.isDitoChannel(normalized)
    DitoNotificationDelegate.shared.application(
      application,
      didReceiveRemoteNotification: userInfo,
      fetchCompletionHandler: completionHandler
    )
    return isDito
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "setDebugMode":
      guard let args = call.arguments as? [String: Any],
            let enabled = args["enabled"] as? Bool else {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "enabled is required and cannot be null",
          details: nil
        ))
        return
      }
      Dito.enableDebugMode(enabled)
      result(nil)
    case "initializeWithApiKey":
      guard let args = call.arguments as? [String: Any],
            let apiKey = args["apiKey"] as? String,
            let bundleId = args["bundleId"] as? String,
            !apiKey.isEmpty, !bundleId.isEmpty else {
        result(FlutterError(
          code: "INVALID_CREDENTIALS",
          message: "apiKey and bundleId are required and cannot be empty",
          details: nil
        ))
        return
      }
      Dito.configure(apiKey: apiKey, bundleId: bundleId)
      result(nil)
    case "initialize":
      guard let args = call.arguments as? [String: Any],
            let appKey = args["appKey"] as? String,
            let appSecret = args["appSecret"] as? String else {
        result(FlutterError(
          code: "INVALID_CREDENTIALS",
          message: "appKey and appSecret are required and cannot be empty",
          details: nil
        ))
        return
      }

      if appKey.isEmpty || appSecret.isEmpty {
        result(FlutterError(
          code: "INVALID_CREDENTIALS",
          message: "appKey and appSecret are required and cannot be empty",
          details: nil
        ))
        return
      }

      Dito.configure(appKey: appKey, appSecret: appSecret)
      result(nil)
    case "identify":
      guard let args = call.arguments as? [String: Any],
            let id = args["id"] as? String else {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "id is required and cannot be empty",
          details: nil
        ))
        return
      }

      if id.isEmpty {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "id is required and cannot be empty",
          details: nil
        ))
        return
      }

      let name = args["name"] as? String
      let email = args["email"] as? String
      let customData = args["customData"] as? [String: Any]

      Dito.identify(
        id: id,
        name: name,
        email: email,
        customData: customData
      )
      result(nil)
    case "track":
      guard let args = call.arguments as? [String: Any],
            let action = args["action"] as? String else {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "action is required and cannot be empty",
          details: nil
        ))
        return
      }

      if action.isEmpty {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "action is required and cannot be empty",
          details: nil
        ))
        return
      }

      let data = args["data"] as? [String: Any]

      Dito.track(
        action: action,
        data: data
      )
      result(nil)
    case "registerDeviceToken":
      guard let args = call.arguments as? [String: Any],
            let token = args["token"] as? String else {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "token is required and cannot be empty",
          details: nil
        ))
        return
      }

      if token.isEmpty {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "token is required and cannot be empty",
          details: nil
        ))
        return
      }

      Dito.registerDevice(token: token)
      result(nil)
    case "unregisterDeviceToken":
      guard let args = call.arguments as? [String: Any],
            let token = args["token"] as? String else {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "token is required and cannot be empty",
          details: nil
        ))
        return
      }

      if token.isEmpty {
        result(FlutterError(
          code: "INVALID_PARAMETERS",
          message: "token is required and cannot be empty",
          details: nil
        ))
        return
      }

      Dito.unregisterDevice(token: token)
      result(nil)
    case "handleNotificationReceived":
      guard let args = call.arguments as? [String: Any] else {
        result(false)
        return
      }
      var userInfo: [AnyHashable: Any] = [:]
      for (key, value) in args where key != "token" {
        userInfo[key] = value
      }
      let token = args["token"] as? String
      let handled = DitoSdkPlugin.didReceiveRemoteNotification(userInfo: userInfo, fcmToken: token)
      result(handled)
    case "handleNotificationClick":
      guard let args = call.arguments as? [String: Any] else {
        result(false)
        return
      }
      var userInfo: [AnyHashable: Any] = [:]
      for (k, v) in args {
        userInfo[k] = v
      }
      // A emissão no stream Dart acontece dentro de didReceiveNotificationClick.
      let handled = DitoSdkPlugin.didReceiveNotificationClick(
        userInfo: userInfo,
        actionIdentifier: args["actionIdentifier"] as? String
      )
      result(handled)
    case "setNotificationOptions":
      let args = call.arguments as? [String: Any] ?? [:]
      let soundResourceName = args["soundResourceName"] as? String
      let options = DitoNotificationOptions(soundName: soundResourceName)
      Dito.setNotificationOptions(options)
      result(nil)
    case "getNotifications":
      let notifications = Dito.shared.getNotifications()
      let maps: [[String: Any]] = notifications.map { info in
        [
          "id": info.id,
          "notificationId": info.notificationId,
          // Sempre vazio: `reference` foi retirado do payload e a atribuição
          // ancora em `user_id`. A chave fica no mapa para não quebrar quem
          // já lê este evento em Dart.
          "reference": "",
          "title": info.title,
          "message": info.message,
          "link": info.link,
          "receivedAt": Int64(info.receivedAt.timeIntervalSince1970 * 1000),
          "isRead": info.isRead,
          "image": info.image,
          "customData": info.customData
        ]
      }
      result(maps)
    case "markNotificationAsRead":
      guard let args = call.arguments as? [String: Any],
            let id = args["id"] as? String else {
        result(FlutterError(code: "INBOX_ERROR", message: "id argument missing", details: nil))
        return
      }
      Dito.shared.markNotificationAsRead(id: id)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    DitoSdkPlugin.notificationEventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    DitoSdkPlugin.notificationEventSink = nil
    return nil
  }
}

File: flutter/android/src/main/kotlin/br/com/dito/ditosdk/DitoSdkPlugin.kt
package br.com.dito.ditosdk

import android.content.Context
import android.os.Handler
import android.os.Looper
import br.com.dito.ditosdk.notification.DitoNotificationOptions
import com.google.firebase.messaging.RemoteMessage
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class DitoSdkPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var notificationEventsChannel: EventChannel
    private var context: Context? = null
    private var debugEnabled: Boolean = false

    companion object {
        private const val NOTIFICATION_EVENTS_CHANNEL = "br.com.dito/notification_events"
        private const val NOTIFICATION_CLICK_EVENT = "notification_click"

        @Volatile
        private var ditoInitialized: Boolean = false

        @Volatile
        private var notificationEventSink: EventChannel.EventSink? = null

        @Volatile
        private var clickDataListenerInstalled: Boolean = false

        /**
         * userInfo do clique que está sendo processado neste exato momento.
         *
         * [NotificationResult] não carrega `log_id`, `notification_name` nem `user_id`, e o
         * stream Dart expõe os três desde sempre. Quando o clique entra pelo method channel
         * temos o mapa completo em mãos, então ele fica aqui para o listener completar o
         * evento. `Dito.notificationClick` invoca o listener de forma síncrona, então não há
         * janela para o valor vazar para outro clique.
         */
        @Volatile
        private var pendingClickUserInfo: Map<String, String>? = null

        @JvmStatic
        fun handleNotification(context: Context, message: RemoteMessage): Boolean {
            if (!isDitoChannel(message)) {
                return false
            }
            ensureDitoInitialized(context)
            processNotificationData(message)
            return true
        }

        /**
         * Encaminha para o stream Dart **todo** clique que o SDK nativo processa.
         *
         * Sem isso o Flutter só vê os cliques que ele mesmo entrega via
         * `handleNotificationClick`, e um toque em botão de ação nunca passa por lá: o
         * `PendingIntent` do botão vai para a `NotificationOpenedActivity` do SDK, então o
         * `FirebaseMessaging.onMessageOpenedApp` — que é o que o app Flutter escuta — não
         * dispara. O listener nativo é o único ponto por onde os dois caminhos passam.
         *
         * Um listener já registrado pelo app host é preservado e chamado antes.
         */
        @JvmStatic
        private fun installClickDataListener() {
            if (clickDataListenerInstalled) return
            clickDataListenerInstalled = true
            val previous = Dito.notificationClickDataListener
            Dito.notificationClickDataListener = { result ->
                previous?.invoke(result)
                emitNotificationClickEvent(result, pendingClickUserInfo ?: emptyMap())
            }
        }

        @JvmStatic
        private fun isDitoChannel(message: RemoteMessage): Boolean {
            val channel = message.data["channel"]
            return channel == "DITO" || channel == "Dito"
        }

        @JvmStatic
        private fun ensureDitoInitialized(context: Context) {
            installClickDataListener()
            if (!ditoInitialized) {
                try {
                    Dito.init(context, null)
                    ditoInitialized = true
                } catch (e: RuntimeException) {
                    android.util.Log.w("DitoSdkPlugin", "Dito SDK not initialized: ${e.message}")
                }
            }
        }

        @JvmStatic
        private fun processNotificationData(message: RemoteMessage) {
            val data = message.data["data"] ?: return
            try {
                val notificationData = parseNotificationData(data)
                if (isValidNotificationData(notificationData)) {
                    sendNotificationRead(notificationData)
                }
            } catch (e: Exception) {
                android.util.Log.e("DitoSdkPlugin", "Error processing notification: ${e.message}")
            }
        }

        @JvmStatic
        private fun parseNotificationData(data: String): NotificationData {
            val jsonData = org.json.JSONObject(data)
            val notificationId = jsonData.optString("notification", "")
            val reference = jsonData.optString("reference", "")
            val logId = jsonData.optString("log_id", "")
            val notificationName = jsonData.optString("notification_name", "")
            val userId = jsonData.optString("user_id", "")
            return NotificationData(notificationId, reference, logId, notificationName, userId)
        }

        @JvmStatic
        private fun isValidNotificationData(data: NotificationData): Boolean {
            return data.notificationId.isNotEmpty() && data.reference.isNotEmpty()
        }

        @JvmStatic
        private fun sendNotificationRead(data: NotificationData) {
            val userInfo = mapOf(
                "notification" to data.notificationId,
                "reference" to data.reference,
                "log_id" to data.logId,
                "notification_name" to data.notificationName,
                "user_id" to data.userId
            )
            Dito.notificationRead(userInfo)
        }

        private data class NotificationData(
            val notificationId: String,
            val reference: String,
            val logId: String,
            val notificationName: String,
            val userId: String
        )

        @JvmStatic
        fun handleNotificationClick(context: Context, userInfo: Map<String, String>): Boolean {
            val channel = userInfo["channel"]
            if (!channel.equals("DITO", ignoreCase = true)) {
                return false
            }
            ensureDitoInitialized(context)
            val normalizedUserInfo = normalizeClickUserInfo(userInfo)
            // A emissão para o Dart acontece no notificationClickDataListener, que cobre
            // este caminho e o do toque em botão. Emitir aqui também duplicaria o evento.
            pendingClickUserInfo = normalizedUserInfo
            try {
                Dito.notificationClick(normalizedUserInfo, null)
            } finally {
                pendingClickUserInfo = null
            }
            return true
        }

        @JvmStatic
        private fun normalizeClickUserInfo(userInfo: Map<String, String>): Map<String, String> {
            val deeplink = userInfo["deeplink"] ?: userInfo["link"] ?: ""
            return userInfo.toMutableMap().apply {
                putIfAbsent("deeplink", deeplink)
            }
        }

        /**
         * Monta o evento do stream Dart a partir do [NotificationResult] do SDK.
         *
         * [userInfo] é o mapa cru do clique quando ele veio pelo method channel, e serve só
         * para completar os campos que o [NotificationResult] não carrega. Quando o clique
         * nasce na notificação nativa esses três campos vêm vazios — o `PendingIntent` não
         * transporta `log_id`/`notification_name`. Antes desta mudança o evento nem existia
         * nesse caminho, então não há regressão.
         */
        @JvmStatic
        private fun emitNotificationClickEvent(
            result: NotificationResult,
            userInfo: Map<String, String>,
        ) {
            val payload: MutableMap<String, Any?> = HashMap()
            payload["type"] = NOTIFICATION_CLICK_EVENT
            payload["deeplink"] = result.deepLink
            payload["notificationId"] = result.notificationId
            payload["reference"] = result.reference
            payload["logId"] = userInfo["log_id"] ?: ""
            payload["notificationName"] = userInfo["notification_name"] ?: ""
            payload["userId"] = userInfo["user_id"] ?: ""
            payload["actionId"] = result.actionId
            payload["actionLabel"] = result.actionLabel
            payload["customData"] = result.customData
            notificationEventSink?.success(payload)
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "br.com.dito/dito_sdk")
        channel.setMethodCallHandler(this)
        notificationEventsChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, NOTIFICATION_EVENTS_CHANNEL)
        notificationEventsChannel.setStreamHandler(this)
        context = flutterPluginBinding.applicationContext
        // O app pode já ter inicializado o SDK por conta própria, e o clique pode chegar
        // antes de qualquer chamada que passe por ensureDitoInitialized.
        installClickDataListener()
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "setDebugMode" -> {
                val enabled = call.argument<Boolean>("enabled")
                if (enabled == null) {
                    result.error(
                        "INVALID_PARAMETERS",
                        "enabled is required and cannot be null",
                        null
                    )
                    return
                }
                debugEnabled = enabled
                Dito.options = Options().apply { debug = enabled }
                result.success(null)
            }
            "initializeWithApiKey" -> {
                val apiKey = call.argument<String>("apiKey")
                val bundleId = call.argument<String>("bundleId")
                if (apiKey.isNullOrEmpty() || bundleId.isNullOrEmpty()) {
                    result.error(
                        "INVALID_CREDENTIALS",
                        "apiKey and bundleId are required and cannot be empty",
                        null
                    )
                    return
                }
                val ctx = context
                if (ctx == null) {
                    result.error(
                        "INITIALIZATION_FAILED",
                        "Context is not available",
                        null
                    )
                    return
                }
                try {
                    Dito.init(ctx, apiKey, "", Options().apply { debug = debugEnabled })
                    ditoInitialized = true
                    result.success(null)
                } catch (e: Exception) {
                    result.error(
                        "INITIALIZATION_FAILED",
                        "Failed to initialize Dito SDK: ${e.message}",
                        null
                    )
                }
            }
            "initialize" -> {
                val appKey = call.argument<String>("appKey")
                val appSecret = call.argument<String>("appSecret")

                if (appKey.isNullOrEmpty() || appSecret.isNullOrEmpty()) {
                    result.error(
                        "INVALID_CREDENTIALS",
                        "appKey and appSecret are required and cannot be empty",
                        null
                    )
                    return
                }

                try {
                    val ctx = context
                    if (ctx == null) {
                        result.error(
                            "INITIALIZATION_FAILED",
                            "Context is not available",
                            null
                        )
                        return
                    }

                    try {
                        val options = Options().apply { debug = debugEnabled }
                        Dito.init(ctx, appKey, appSecret, options)
                        ditoInitialized = true
                        result.success(null)
                    } catch (e: RuntimeException) {
                        if (e.message?.contains("API_KEY e API_SECRET no AndroidManifest") == true) {
                            result.error(
                                "INITIALIZATION_FAILED",
                                "Dito SDK requires API_KEY and API_SECRET to be configured in AndroidManifest.xml. Please add them to your app's AndroidManifest.xml file.",
                                null
                            )
                        } else {
                            throw e
                        }
                    }
                } catch (e: Exception) {
                    result.error(
                        "INITIALIZATION_FAILED",
                        "Failed to initialize Dito SDK: ${e.message}",
                        null
                    )
                }
            }
            "identify" -> {
                val id = call.argument<String>("id")
                val name = call.argument<String>("name")
                val email = call.argument<String>("email")
                val customData = call.argument<Map<String, Any>>("customData")

                if (id.isNullOrEmpty()) {
                    result.error(
                        "INVALID_PARAMETERS",
                        "id is required and cannot be empty",
                        null
                    )
                    return
                }

                try {
                    Dito.identify(
                        id = id,
                        name = name,
                        email = email,
                        customData = customData
                    )
                    result.success(null)
                } catch (e: Exception) {
                    result.error(
                        "NETWORK_ERROR",
                        "Failed to identify user: ${e.message}",
                        null
                    )
                }
            }
            "track" -> {
                val action = call.argument<String>("action")
                val data = call.argument<Map<String, Any>>("data")

                if (action.isNullOrEmpty()) {
                    result.error(
                        "INVALID_PARAMETERS",
                        "action is required and cannot be empty",
                        null
                    )
                    return
                }

                try {
                    Dito.track(
                        action = action,
                        data = data
                    )
                    result.success(null)
                } catch (e: Exception) {
                    result.error(
                        "NETWORK_ERROR",
                        "Failed to track event: ${e.message}",
                        null
                    )
                }
            }
            "registerDeviceToken" -> {
                val token = call.argument<String>("token")

                if (token.isNullOrEmpty()) {
                    result.error(
                        "INVALID_PARAMETERS",
                        "token is required and cannot be empty",
                        null
                    )
                    return
                }

                try {
                    Dito.registerDevice(token)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(
                        "NETWORK_ERROR",
                        "Failed to register device token: ${e.message}",
                        null
                    )
                }
            }
            "unregisterDeviceToken" -> {
                val token = call.argument<String>("token")

                if (token.isNullOrEmpty()) {
                    result.error(
                        "INVALID_PARAMETERS",
                        "token is required and cannot be empty",
                        null
                    )
                    return
                }

                try {
                    Dito.unregisterDevice(token)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(
                        "NETWORK_ERROR",
                        "Failed to unregister device token: ${e.message}",
                        null
                    )
                }
            }
            "handleNotificationClick" -> {
                val ctx = context
                if (ctx == null) {
                    result.success(false)
                    return
                }
                val args = call.arguments as? Map<*, *>
                if (args == null) {
                    result.success(false)
                    return
                }

                val userInfo: MutableMap<String, String> = HashMap()
                for ((key, value) in args) {
                    if (key == null) continue
                    userInfo[key.toString()] = value?.toString() ?: ""
                }
                result.success(handleNotificationClick(ctx, userInfo))
            }
            "setNotificationOptions" -> {
                val args = call.arguments as? Map<*, *> ?: run {
                    result.success(null)
                    return
                }
                val options = DitoNotificationOptions(
                    smallIconResId = (args["smallIconResId"] as? Number)?.toInt(),
                    largeIconResId = (args["largeIconResId"] as? Number)?.toInt(),
                    soundResourceName = args["soundResourceName"] as? String,
                    accentColor = (args["accentColor"] as? Number)?.toInt()
                )
                Dito.setNotificationOptions(options)
                result.success(null)
            }
            "getNotifications" -> {
                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        val records = Dito.getNotifications()
                        val maps = records.map { info ->
                            mapOf(
                                "id" to info.id,
                                "notificationId" to info.notificationId,
                                "reference" to info.reference,
                                "title" to info.title,
                                "message" to info.message,
                                "link" to info.link,
                                "receivedAt" to info.receivedAt,
                                "isRead" to info.isRead,
                                "image" to info.image,
                                "customData" to info.customData
                            )
                        }
                        Handler(Looper.getMainLooper()).post { result.success(maps) }
                    } catch (e: Exception) {
                        Handler(Looper.getMainLooper()).post { result.error("INBOX_ERROR", e.message, null) }
                    }
                }
            }
            "markNotificationAsRead" -> {
                val args = call.arguments as? Map<*, *>
                val id = args?.get("id") as? String
                if (id == null) {
                    result.error("INBOX_ERROR", "id argument missing", null)
                    return
                }
                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        Dito.markNotificationAsRead(id)
                        Handler(Looper.getMainLooper()).post { result.success(null) }
                    } catch (e: Exception) {
                        Handler(Looper.getMainLooper()).post { result.error("INBOX_ERROR", e.message, null) }
                    }
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        notificationEventSink = events
    }

    override fun onCancel(arguments: Any?) {
        notificationEventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        notificationEventsChannel.setStreamHandler(null)
        notificationEventSink = null
        context = null
    }
}
